base/abstractarraymath.jl

```diff
@@ -330,89 +308,171 @@ julia> repeat([1 2; 3 4], inner=(2, 1), outer=(1, 3))
+    a1 = first(a)
+    b1 = first(b)
+    a_tail, b_tail = pad(Base.tail(a), Base.tail(b))
+    tuple(a1, a_tail...), tuple(b1, b_tail...)
+end
+
+function resolve(arr::AbstractArray{<:Any, N}, inner::NTuple{N, Any}, outer::NTuple{N,Any}) where {N}
+    arr, inner, outer
+end
+function resolve(arr, inner, outer)
+    dims, inner, outer = pad(size(arr), to_tuple(inner), to_tuple(outer))
+    reshape(arr, dims), inner, outer
+end
+function resolve(arr, inner::Nothing, outer::Nothing)
+    return arr, inner, outer
+end
+function resolve(arr, inner::Nothing, outer)
+    dims, outer = pad(size(arr), to_tuple(outer))
+    reshape(arr, dims), inner, outer
+end
+function resolve(arr, inner, outer::Nothing)
+    dims, inner = pad(size(arr), to_tuple(inner))
+    reshape(arr, dims), inner, outer
+end
+
+function check(arr, inner, outer)
+    if inner != nothing
+        if any(<(0), inner)
+            msg = """
+            Error calling repeat. all(inner . >= 0) must hold. Got
+            inner = $inner
+            """
+            throw(ArgumentError(msg))
+        end
+        if length(inner) < ndims(arr)
+            msg = """
+            Error calling repeat. length(inner) >= ndims(arr) must hold. Got
+            inner = $inner
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
     end
+    if outer != nothing
+        if any(<(0), outer)
+            msg = """
+            Error calling repeat. all(outer . >= 0) must hold. Got
+            outer = $outer
+            """
+            throw(ArgumentError(msg))
+        end
+        if (length(outer) < ndims(arr)) && (inner != nothing)
+            msg = """
+            Error calling repeat. length(outer) >= ndims(arr) must hold. Got
+            outer = $outer
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
+    end
+end
 
-    # fill the first inner block
-    if all(isequal(1), inner)
-        idxs = (axes(A)..., ntuple(n->OneTo(1), ndims(R)-ndims(A))...) # keep dimension consistent
-        R[idxs...] = A
-    else
-        inner_indices = [1:n for n in inner]
-        for c in CartesianIndices(axes(A))
-            for i in 1:ndims(A)
-                n = inner[i]
-                inner_indices[i] = (1:n) .+ ((c[i] - 1) * n)
-            end
-            fill!(view(R, inner_indices...), A[c])
+repeat_inner_outer(arr, inner::Nothing, outer::Nothing) = arr
+repeat_inner_outer(arr, ::Nothing, outer) = repeat_outer(arr, outer)
+repeat_inner_outer(arr, inner, ::Nothing) = repeat_inner(arr, inner)
+repeat_inner_outer(arr, inner, outer) = repeat_outer(repeat_inner(arr, inner), outer)
+
+function repeat_outer(a::AbstractMatrix, (m,n)::NTuple{2, Any})
+    o, p = size(a,1), size(a,2)
+    b = similar(a, o*m, p*n)
+    for j=1:n
+        d = (j-1)*p+1
+        R = d:d+p-1
+        for i=1:m
+            c = (i-1)*o+1
+            @inbounds b[c:c+o-1, R] = a
```

Comment:
Just for my own reference (since GitHub isn't aligning these diffs helpfully), this is nearly exactly the status quo, but with this extra `@inbounds`.  Looks good.